### PR TITLE
post: *How To Securely Share Information*

### PR DESCRIPTION
**Pull request by thejohncrafter**
A post where I describe what I've done with "secure aggregation"
* [x] Proof read
* [x] Extra part
* [x] Correct spelling (fixed or at least I hope so)

### CODE REVIEW

File: _posts/2022-02-12-how-to-securely-share-information.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+Concretely, let's say we have four users, Alice, Beatrice, Christine and Dominique.
+Let's say their preferences can be encoded over 4 bits[^scale], and that the encoded preference is never null.
+Alice and her friends will build an 8 bits long message that will contain their preferences.
+We use 16 bits because the final message will contain four slots, one for each individual message.
+Alice's preferences are $${\mathtt{1001}}$$, Beatrice's are $${\mathtt{0110}}$$,
+Christine prefers $${\mathtt{1010}}$$ and Dominique votes $${\mathtt{0101}}$$ (we can use any other non-null encoding).
+
+![Our users communicating with Mangaki](/public/img/persons.png)
+
+Alice, Beatrice, Christine and Dominique just need to know how where to put their individual preferences in the final vector.
```
Comment (jilljenn) [inline _posts/2022-02-12-how-to-securely-share-information.md]: just need to know how*

File: _posts/2022-02-12-how-to-securely-share-information.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+the result can be safely published without compromising the users' privacy.
+
+[^scale]: 4 users and 4 bits is ridiculously small, but it is easier to picture, and the example scales well !
+
+[^binaryops]: We said that bit strings are vectors, and we used the _exclusive or_ operation $$\wedge$$.
+  Mathematically, we were just writing about vectors over the two element field.
+
+## More detail
+
+The previous example is, of course, degenerate, because there are only two users, and our strategy to find slots for each user is too expansive.
```
Comment (jilljenn) [inline _posts/2022-02-12-how-to-securely-share-information.md]: four*

Comment (jilljenn) [reply _posts/2022-02-12-how-to-securely-share-information.md]: expensive*

File: _posts/2022-02-12-how-to-securely-share-information.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+The first possibility is that, since each user knows where their own preferences lie in the collected data,
+they can download the whole model and perform the dot products themselves without the server knowing
+which of the collected preferences belongs to the user.
+This implies no more leakage of data than what we currently have.
+
+We could also have the server publish the data of every anonymized user (publish the dot products directly), which amounts to the same,
+this just changes where the computation happens but not the security model.
+
+We could also have the server publish just the embedding of every anime, and the users train their own local model.
+With this technique, we loose a little accuracy on the predictions, but we gain several benefits:
```
Comment (jilljenn) [inline _posts/2022-02-12-how-to-securely-share-information.md]: lose

**Comment by jilljenn** [comment]
- I understood you use Shamir secret sharing (possibly Lagrange polynomials) while for secure computation we use ElGamal / Paillier (which needs all clients to be there for decryption). Maybe you could have mentioned it in the post.
- I don't fully understand why "finding a slot for a user" is relevant; I don't see anything like this in the reference article. (Maybe it refers to the masks mentioned in the article?)
Thanks! LGTM

**Comment by thejohncrafter** [comment]
To answer your two points:
 * Yes, I use among other things Shamir Secret Sharing. I also use some other encryption schemes like authenticated encryption and key agreement, behind the scenes the libraries I use are built with curve25519 cryptography (from what I remember). I think I can mention that in an "implementation details" section, where I write the libraries I use (which would be good for the discoverability of these libraries, btw).
  But I don't feel like it is a that relevant to compare what I write about in this post with the secure computation technique (unless I spend several paragraphs on that), since secure aggregation and secure computation are entirely different, and comparing just the cryptographic primitives that are used in these would be misleading as the reader would probably think our implementations just boil down to trivial uses of widely known cryptography while it clearly does not.
 * I write about "finding a slot" because what I want to do is to get a list of all the preferences of my users without knowing which preference belongs to whom. The basic idea is that with _secure aggregation_, I can perform ``XOR`` on bitstrings, knowing the result without knowing the operands. So, to make a comparison, it's like if my users were writing on a piece of paper with some magic ink, and then revealing the whole thing. They can't know who wrote where because they "didn't watch". But the problem is that they can't know if they are writing at a place where someone else wrote. That's where "finding a slot" comes. It's just about having somewhere to write without being overwritten.
   So no, it is not in the original paper, it is a problem that arises with what I do with the paper. The blog post is actually divided in two parts: in the first, I write about what the paper enables us to do, and in the second, I build on top of it to securely collect data. The second kind of like an informal extension to the original paper. I guess that's not clear enough in the post I wrote, maybe I should add a new paragraph to state this very clearly ?

**Comment by RaitoBezarius** [comment]
cc @jilljenn

**Comment by jilljenn** [comment]
OK j'avais déjà LGTM et je t'avais dit que tu pouvais review @RaitoBezarius :) Good job

**Pull request merged**